doc/developer/design/20231204_query_lifecycle_events_logging.md

```diff
@@ -0,0 +1,149 @@
+
+### Time to result delivery
+
+I originally attempted to report the time it took to deliver the
+results over the network. However, after some thought, I realized
+there's no good way to do this. We could try to detect when all the
+bytes we've been sent have been `ACK`ed by repeatedly polling a
+Linux system call to get this information. However, this would not
+measure just the time to send the data, but also the time for the ACK
+to be returned to us.
+
+## Solution Proposal
+
+Create a collection called `mz_internal.mz_statement_lifecycle_history`
+with the schema `(statement_id uuid NOT NULL, event_type text NOT NULL,
+occurred_at timestamp with time zone NOT NULL)`.
+
+The `event_type` can be one of the following:
+
+| type                         | description                                                                                                              |
+|------------------------------|--------------------------------------------------------------------------------------------------------------------------|
+| `execution-began`            | Corresponds to today's `mz_statement_execution_history.began_at`                                                         |
+| `storage-dependencies-ready` | When all transitive source dependencies advanced to the timestamp. If there are none, same as `execution_began`          |
+| `compute-dependencies-ready` | When all compute dependencies (MVs and Indexes) advanced to the timestamp. If there are none, same as `execution_began`. |
+| `execution-finished`         | Corresponds to today's `mz_statement_execution_history.finished_at`                                                      |
+| `last-row-returned`          | The time at which all rows were sent and we were ready to receive new statements from its session.                       |
+
+When each event is generated, the coordinator will insert it into a
+buffer which is periodically dumped to storage, similarly to the
+statement logging events that exist today. Since the amount of data is
+proportional to the number of statements, we don't envision that this
+will result in new scalability concerns beyond those which already
+exist for the statement log.
+
+In the following subsections we will explain how to collect each of
+the pieces of data.
+
+### `execution-began`
+
+We already collect this.
+
+### `storage-dependencies-ready` / `compute-dependencies-ready`
+
+Introduce a mechanism for the sub-controllers (Storage and Compute
+controllers) to return all frontier updates to the overall
+Controller. When installing a query with the Controller, the
+Coordinator will also inform it of one or more "watch sets" of
+IDs. Concretely, the Controller will gain a function like:
+
+``` rust
+fn add_watch_set(&mut self, id: Box<Any>, global_ids: Vec<GlobalId>, ts: Timestamp);
+```
+
+The Controller will then be expected to return responses of the form:
+
+``` rust
+WatchSetFulfilled {
+    id: Box<Any>,
+}
+```
```

Comment:
Is this returned immediately from `add_watch_set` or is this sent to the Coordinator at a later time via some channel? Could you either update the the signature of `add_watch_set` or add some info about how the controller sends this information to the Coordinator.